### PR TITLE
Clear local transactions that are known by backend

### PR DESCRIPTION
**Pull request by everdimension**

### CODE REVIEW

File: src/background/transactions/TransactionService.ts [hunk 1]
```diff
@@ -52,9 +67,88 @@ export class TransactionService {
...
+    address,
+    fromNonce,
+    chainId,
+  }: {
+    address: string;
+    fromNonce: number;
+    chainId: ChainId;
+  }) {
+    const transactions = await this.transactionsStore.getSavedState();
+    const candidates = transactions.filter((item) => {
```
Comment (zerts) [inline src/background/transactions/TransactionService.ts:108]: This is indeed an optimisation, but it looks like we can pass all transactions to `bulkDeleteTransactionsByHash` without significant difference in speed

File: src/background/transactions/TransactionService.ts [hunk 2]
```diff
@@ -52,9 +67,88 @@ export class TransactionService {
...
+    for (const item of transactions) {
+      const chainId = normalizeChainId(item.transaction.chainId);
+      const key = `${item.transaction.from}:${chainId}` as const;
+      map.set(key, { hash: item.hash, timestamp: item.timestamp });
+    }
+
+    for (const [key, { hash, timestamp }] of map.entries()) {
+      const [address, chainIdStr] = key.split(':');
+      const chainId = chainIdStr as ChainId;
+      const network = await getNetworkByChainId(chainId);
```
Comment (zerts) [inline src/background/transactions/TransactionService.ts:137]: Why not using networksStore here?

**Review by zerts — APPROVED** [review]

**Comment by zerts** [comment]
Just some questions here. Nothing to block the merge

File: src/background/transactions/TransactionService.ts [hunk 3]
```diff
@@ -52,9 +67,88 @@ export class TransactionService {
...
+    const transactions = this.transactionsStore.getState();
+    const pending = getPendingTransactions(transactions);
     this.transactionsPoller.add(pending.map(toPollingObj));
     this.addListeners();
+    if (transactions.length) {
+      this.startPurgeInterval({ leading: true });
+    }
+  }
+
+  private startPurgeInterval({ leading } = { leading: false }) {
```
[the remark this answers is not in the document]
Comment (everdimension) [reply src/background/transactions/TransactionService.ts:79]: Thx, I moved private methods before public ones

File: src/background/transactions/TransactionService.ts (hunk 1 above)
Comment (zerts) [reply src/background/transactions/TransactionService.ts:108]: Incorrect comment

File: src/background/transactions/TransactionService.ts (hunk 2 above)
Comment (zerts) [reply src/background/transactions/TransactionService.ts:137]: Ok then. This will just remove an extra call for a chain info probably

Comment (everdimension) [reply src/background/transactions/TransactionService.ts:137]: But it seems that we want fresh data here

File: src/background/transactions/TransactionService.ts [hunk 4]
```diff
@@ -87,6 +181,8 @@ export class TransactionService {
             draft.push(newItem);
           })
         );
+        this.startPurgeInterval();
```
[the remark this answers is not in the document]
Comment (everdimension) [reply src/background/transactions/TransactionService.ts:184]: Yeah, that will effectively behave the same. I guess it just feels nice not to run code and schedule an interval when it's known ahead of time that it's not needed

File: src/background/transactions/TransactionService.ts [hunk 5]
```diff
@@ -41,20 +57,126 @@ export class TransactionService {
   private transactionsStore: TransactionsStore;
   private transactionsPoller: TransactionsPoller;
 
+  static ALARM_NAME = 'TransactionService:performPurgeCheck';
```
Comment (zerts) [inline src/background/transactions/TransactionService.ts:60]: May be it is better to name this `ALARM_NAME` more explicit?

File: src/background/transactions/TransactionService.ts [hunk 6]
```diff
@@ -41,20 +57,126 @@ export class TransactionService {
...
+      const key = `${item.transaction.from}:${chainId}` as const;
+      map.set(key, { hash: item.hash, timestamp: item.timestamp });
+    }
+
+    for (const [key, { hash, timestamp }] of map.entries()) {
+      const [address, chainIdStr] = key.split(':');
+      const chainId = chainIdStr as ChainId;
+      const network = await getNetworkByChainId(chainId);
+      if (network?.supports_actions) {
+        const knownNonce = await getLatestNonceKnownByBackend({
```
Comment (zerts) [inline src/background/transactions/TransactionService.ts:167]: Will we make a call for all actions one by one here?
Should we somehow exclude removed hashes from the `map`?

**Review by zerts — APPROVED** [review]

File: src/background/transactions/TransactionService.ts (hunk 5 above)
Comment (everdimension) [reply src/background/transactions/TransactionService.ts:60]: More explicit than `'TransactionService:performPurgeCheck'`? :)

**Comment by abalyasnikov** [comment]
qa verified

**Pull request merged**